crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -1128,32 +1122,77 @@ impl Account {
+    /// This tries to decrypt an Olm message using all the sessions we share
     /// with the given sender.
-    async fn decrypt_with_existing_sessions(
+    async fn decrypt_normal_olm_message(
         store: &Store,
+        sender: &UserId,
         sender_key: Curve25519PublicKey,
         message: &OlmMessage,
-    ) -> OlmResult<Option<(Session, String)>> {
-        let s = store.get_sessions(&sender_key.to_base64()).await?;
+    ) -> OlmResult<(SessionType, String)> {
+        let OlmMessage::Normal(_) = message else {
+            panic!("decrypt_normal_olm_message called with a prekey message");
+        };
+
+        let existing_sessions = store.get_sessions(&sender_key.to_base64()).await?;
+        if let Some(ref sessions) = existing_sessions {
+            let sessions: &mut Vec<Session> = &mut *sessions.lock().await;
+            // Try to decrypt the message using each Session we share with the
+            // given curve25519 sender key.
+            for session in sessions {
+                if let Ok(p) = session.decrypt(message).await {
+                    // success!
+                    return Ok((SessionType::Existing(session.clone()), p));
+                } else {
+                    // An error here is completely normal, after all we don't know
+                    // which session was used to encrypt a message. We will log a
+                    // warning if no session was able to decrypt the message.
+                    continue;
+                }
+            }
+        }
+
+        let session_ids = if let Some(sessions) = existing_sessions {
+            sessions.lock().await.iter().map(|s| s.session_id().to_owned()).collect()
+        } else {
+            vec![]
+        };
+
+        warn!(?session_ids, "Failed to decrypt a non-pre-key message with all available sessions");
+        return Err(OlmError::SessionWedged(sender.to_owned(), sender_key));
+    }
 
-        let Some(sessions) = s else {
-            // We don't have any existing sessions, return early.
-            return Ok(None);
+    /// Try to decrypt a prekey Olm message.
+    ///
+    /// If we already know about the session corresponding to the message, use
+    /// it to decrypt the message. Otherwise, create a new Olm session.
+    async fn decrypt_prekey_olm_message(
+        &self,
+        store: &Store,
+        sender: &UserId,
+        sender_key: Curve25519PublicKey,
+        message: &OlmMessage,
+    ) -> OlmResult<(SessionType, String)> {
+        let OlmMessage::PreKey(ref prekey_message) = message else {
+            panic!("decrypt_prekey_olm_message called with a non-prekey message");
         };
 
-        let mut decrypted: Option<(Session, String)> = None;
+        // First try to decrypt using an existing session.
+        let existing_sessions = store.get_sessions(&sender_key.to_base64()).await?;
+        if let Some(ref sessions) = existing_sessions.as_ref() {
```

Comment:
Please no `ref` patterns. I also don't see `existing_sessions` used again, so no extra `&` or `.as_ref()` should be needed (which should then also eliminate the `.clone()` below)

```suggestion
        if let Some(sessions) = existing_sessions {
```

Reply:
Yeah I can certainly clean this up.

I think we still need to `clone` the session though, so we can wrap it in the `SessionType` return type. Note that the previous version of this code also had a clone: https://github.com/matrix-org/matrix-rust-sdk/pull/2751/files#diff-909c5de6771a2992465b595c86bbc6baacc92acd231a79cee788f16c8d4fcbd5L1153